Cython/Compiler/Code.py

```diff
@@ -1423,23 +1423,33 @@ def new_string_const_cname(self, bytes_value):
         value = bytes_value.decode('ASCII', 'ignore')
         return self.new_const_cname(value=value)
 
-    def new_num_const_cname(self, value, py_type):
+    def make_unique_const_cname(self, format_str):  # type: (str) -> str
+        used = self.const_cnames_used
+        cname = format_str % ""
+        counter = 0
+        while used.get(cname):
+            counter += 1
+            cname = format_str % counter
+        used[cname] = 1
+        return cname
+
+    def new_num_const_cname(self, value, py_type):  # type: (str, str) -> str
         if py_type == 'long':
             value += 'L'
             py_type = 'int'
         prefix = Naming.interned_prefixes[py_type]
-        cname = "%s%s" % (prefix, value)
-        cname = cname.replace('+', '_').replace('-', 'neg_').replace('.', '_')
+
+        value = value.replace('.', '_').replace('+', '_').replace('-', 'neg_')
+        if len(value) > 42:
+            cname = self.make_unique_const_cname(
+                prefix + "large%s_" + value[:21] + "_XXX_" + value[-21:])
```

Comment:
Let's make the split less "loud". Also, let's avoid making the identifiers longer when "shortening" them.
```suggestion
                prefix + "large%s_" + value[:18] + "_xxx_" + value[-18:])
```